File: blackjack/java/outcomes.java
class outcomes {
    public static int partitions(int[] cards, int subtotal)
    {
	int m = 0;
	// Hit
	for (int i = 0; i <= 9; i++) {
	    if (cards[i]>0) {
		int total = subtotal+i+1;
		if (total < 21) {
		    // Stand
		    m += 1;
		    // Hit again
		    cards[i] -= 1;
		    m += partitions(cards, total);
		    cards[i] += 1;
		} else if (total==21) {
		    // Stand; hit again is an automatic bust
		    m += 1;
		}
	    }
	}
	return m;
    }

    public static void main(String[] args)
    {
	int[] deck = new int[]{4,4,4,4,4,4,4,4,4,16};
	int d = 0;

	for (int i = 0; i <= 9; i++) {
	    // Dealer showing
	    deck[i] -= 1;
	    int p = 0;
	    for (int j = 0; j <= 9; j++) {
		deck[j] -= 1;
		p += partitions(deck, j+1);
		deck[j] += 1;
	    }
	    System.out.println("Dealer showing "+i+" partitions = "+p);
	    d += p;
	    deck[i] += 1;
	}
	System.out.println("Total partitions = "+d);
    }
}

File: blackjack/vb/outcomes.vb
module outcomes

function partitions(cards() as integer, subtotal as integer) as integer

    dim i as integer
    dim m as integer
    dim total as integer
    m = 0
    ' Hit
    for i = 0 to 9
        if (cards(i)>0) then
            total = subtotal+i+1
            if (total < 21) then
                ' Stand
                m += 1
                ' Hit again
		cards(i) -= 1
                m = m+partitions(cards, total)
                cards(i) += 1
            elseif (total=21) then
                ' Stand; hit again is an automatic bust
                m += 1
	    end if
        end if
    next
    return m
end function

sub main()

dim deck = new integer(9) {4,4,4,4,4,4,4,4,4,16}
dim d as integer
dim i as integer
dim p as integer
dim j as integer
d = 0

for i = 0 to 9
    ' Dealer showing
    deck(i) -= 1
    p = 0
    for j = 0 to 9
        deck(j) -= 1
        p += partitions(deck, j+1)
        deck(j) += 1
    next
    console.writeline("Dealer showing {0} partitions = {1}", i, p)
    d += p
    deck(i) += 1
next

console.writeline("Total partitions = {0}", d)

end sub

end module